lib/SILOptimizer/Analysis/EscapeAnalysis.cpp

```diff
@@ -575,32 +827,111 @@ getUsePoints(CGNode *Node, llvm::SmallVectorImpl<SILNode *> &UsePoints) {
-  // So this should be fast.
-  llvm::SmallVector<CGNode *, 8> WorkList;
-  WorkList.push_back(From);
-  From->isInWorkList = true;
-  for (unsigned Idx = 0; Idx < WorkList.size(); ++Idx) {
-    CGNode *Reachable = WorkList[Idx];
-    if (Reachable == To) {
-      clearWorkListFlags(WorkList);
-      return true;
+// Traverse backward from startNode and return true if \p visitor did not halt
+// traversal..
+//
+// The graph may have cycles.
+template <typename CGPredVisitor>
+bool EscapeAnalysis::ConnectionGraph::backwardTraverse(
+    CGNode *startNode, CGPredVisitor &&visitor) {
+  CGNodeWorklist worklist(this);
+  worklist.push(startNode);
+
+  for (unsigned idx = 0; idx < worklist.size(); ++idx) {
+    CGNode *reachingNode = worklist[idx];
+
+    for (Predecessor pred : reachingNode->Preds) {
+      switch (visitor(pred)) {
+      case Traversal::Follow: {
+        CGNode *predNode = pred.getPointer();
+        worklist.tryPush(predNode);
+        break;
+      }
+      case Traversal::Backtrack:
+        break;
+      case Traversal::Halt:
+        return false;
+      }
     }
-    for (Predecessor Pred : Reachable->Preds) {
-      CGNode *PredNode = Pred.getPointer();
-      if (!PredNode->isInWorkList) {
-        PredNode->isInWorkList = true;
-        WorkList.push_back(PredNode);
+  }
+  return true;
+}
+
+// Traverse forward from startNode, following defer edges and return true if \p
+// visitor did not halt traversal.
+//
+// The graph may have cycles.
+template <typename CGNodeVisitor>
+bool EscapeAnalysis::ConnectionGraph::forwardTraverseDefer(
+    CGNode *startNode, CGNodeVisitor &&visitor) {
+  CGNodeWorklist worklist(this);
+  worklist.push(startNode);
+
+  for (unsigned idx = 0; idx < worklist.size(); ++idx) {
+    CGNode *reachableNode = worklist[idx];
+
+    for (CGNode *deferNode : reachableNode->defersTo) {
+      switch (visitor(deferNode)) {
+      case Traversal::Follow:
+        worklist.tryPush(deferNode);
+        break;
+      case Traversal::Backtrack:
+        break;
+      case Traversal::Halt:
+        return false;
       }
     }
   }
-  clearWorkListFlags(WorkList);
-  return false;
+  return true;
 }
 
+// Follow transitive pointsTo edges from \p startNode.
+// This path may not contain cycles.
+template <typename CGNodeVisitor>
+bool EscapeAnalysis::ConnectionGraph::forwardTraversePointsToEdges(
+    CGNode *startNode, CGNodeVisitor &&visitor) {
+  CGNodeWorklist visitedNodes(this);
+  CGNode *pointer = startNode;
+  while ((pointer = pointer->getPointsToEdge())) {
+    if (!visitedNodes.tryPush(pointer))
+      return true;
+
+    if (!visitor(pointer)) {
+      return false;
+    }
+  }
+  return true;
+}
+
+// Returns the last content node in the chain of pointsTo edges pointed to by
+// this node.
+//
+// If this node has no content, return the node itself. If has content or is
+// already a content follow the chain of pointsTo edges and return the last
+// content node in the chain.
```

Comment:
What if the points-to subgraph contains a loop? Should it return null in such a case (currently it returns an arbitrary node of the loop)?
This makes me wonder if this function is useful at all? Do we need the end of the point-to-chain _in case it's not a loop_?

Reply:
The purpose of `getPointsToEnd` was to get the content node with the weakest constraint (i.e. the one most likely to escape). So when there's a loop, any node in the loop was ok. But it's not an obviously useful standalone API so I'll remove it from this PR.